tempesta_fw/cfg.c

```diff
@@ -0,0 +1,1563 @@
+	         || !strcasecmp(in_str, "off")
+	         || !strcasecmp(in_str, "no")
+	         || !strcasecmp(in_str, "false")
+	         || !strcasecmp(in_str, "disable");
+
+	BUG_ON(is_true && is_false);
+	if (!is_true && !is_false) {
+		TFW_ERR("invalid boolean value: '%s'\n", in_str);
+		return -EINVAL;
+	}
+
+	*dest_bool = is_true;
+	return 0;
+}
+EXPORT_SYMBOL(tfw_cfg_set_bool);
+
+/**
+ * Detect integer base and strip 0x and 0b prefixes from the string.
+ *
+ * The custom function is written because the kstrtox() treats leading zeros as
+ * the octal base. That may cause an unexpected effect when you specify "010" in
+ * the configuration and get 8 instead of 10. We want to avoid that.
+ *
+ * As a bonus, we have the "0b" support here. This may be handy for specifying
+ * some masks and bit strings in the configuration.
+ */
+static int
+detect_base(const char **pos)
+{
+	const char *str = *pos;
+	size_t len = strlen(str);
+
+	if (!len)
+		return 0;
+
+	if (len > 2 && str[0] == '0' && isalpha(str[1])) {
+		char c = tolower(str[1]);
+
+		(*pos) += 2;
+
+		if (c == 'x')
+			return 16;
+		else if (c == 'b')
+			return 2;
+		else
+			return 0;
+	}
+
+	return 10;
+}
+
+int
+tfw_cfg_set_int(TfwCfgSpec *cs, TfwCfgEntry *e)
+{
+	int base, r, val;
+	int *dest_int;
+	const char *in_str;
+
+
+	BUG_ON(!cs->dest);
+	r = tfw_cfg_check_single_val(e);
+	if (r)
+		goto err;
+
+	in_str = e->vals[0];
+	base = detect_base(&in_str);
+	if (!base)
+		goto err;
+
+	r = kstrtoint(in_str, base, &val);
+	if (r)
+		goto err;
+
+	if (cs->spec_ext) {
+		TfwCfgSpecInt *cse = cs->spec_ext;
+
+		if (cse->is_multiple_of && (val % cse->is_multiple_of)) {
+			TFW_ERR("the value of '%s' is not a multiple of %d\n",
+				in_str ,cse->is_multiple_of);
+			goto err;
+		}
+
+		if (cse->range.min != cse->range.max &&
+		    (val < cse->range.min || val > cse->range.max)) {
+			TFW_ERR("the value of '%s' is out of range: %ld, %ld\n",
+				in_str, cse->range.min, cse->range.max);
+			goto err;
+		}
+	}
+
+	dest_int = cs->dest;
+	*dest_int = val;
+	return 0;
+err:
+	TFW_ERR("can't parse integer");
+	return -EINVAL;
+}
+EXPORT_SYMBOL(tfw_cfg_set_int);
+
+int
+tfw_cfg_set_str(TfwCfgSpec *cs, TfwCfgEntry *e)
+{
+	const char *in_str = e->vals[0];
+	const char **dest_str = cs->dest;
+	TfwCfgSpecStr *cse = cs->spec_ext;
+	int r, len, len_min, len_max;
+
+	r = tfw_cfg_check_single_val(e);
+	if (r)
+		return r;
+
+	BUG_ON(!dest_str);
+	BUG_ON(!cs);
+	BUG_ON(!cse->buf.buf || !cse->buf.size); /* TODO: dynamic allocation. */
+
+	len = strlen(in_str);
+	if (len >= cse->buf.size) {
+		TFW_ERR("the string is too long: '%s'\n", in_str);
+		return -EINVAL;
+	}
+
+	len_min = cse->len_range.min;
+	len_max = cse->len_range.max;
+	if (len_min != len_max && (len < len_min || len > len_max)) {
+		TFW_ERR("the string length (%d) is out of valid range (%d, %d):"
+			" '%s'\n", len, len_min, len_max, in_str);
+		return -EINVAL;
+	}
+
+	memcpy(cse->buf.buf, in_str, len + 1);
+	*dest_str = cse->buf.buf;
```

Comment:
It seems something wrong is going with dest_str. Why do you need the variable and why it's a double pointer? Why not just to call ```cse->dest = cse->buf.buf``` here?